src/analyses/localTraceSemantics/locTraceAnalysis.ml

```diff
@@ -0,0 +1,1388 @@
+                            in
+                            let unlockedNode = {programPoint=ctx.node;sigma=newSigma;id=(idGenerator#getID assignedNode unlockingLabel ctx.node newSigma tid ls);tid=tid;lockSet=ls}
+                            in
+                            let unlockingEdge = (assignedNode, unlockingLabel, unlockedNode)
+                            in
+                            let unlockingGraph = LocalTraces.extend_by_gEdge assignedGraph unlockingEdge
+                            in
+                            let ctxGlobalTid = ctx.global myTmp
+                            in
+                            ctx.sideg myTmp (D.add unlockingGraph ctxGlobalTid);
+                            unlockingGraph::resultGraphList
+
+                          ) [] lockedGraphList
+                        in
+                        graphList@outerGraphList
+                      ) else(
+                        let (myEdge:(node * CustomEdge.t * node)) =
+                          (lastNode, (Assign(lval,newExp)),{programPoint=ctx.node;sigma=newSigma;id=(idGenerator#getID lastNode (Assign(lval,newExp)) ctx.node newSigma tid ls);tid=tid;lockSet=ls})
+                        in
+                        print_string ("assignment succeeded so we add the edge "^(LocalTraces.show_edge myEdge)^"\n");
+                        (LocalTraces.extend_by_gEdge graph_outter myEdge)::outerGraphList )
+                    ) [] sigmaList))@resultList_outter
+         ) [] someTmp
+       in
+       resultList
+                   | _ -> Printf.printf "This type of assignment is not supported in assign_on_node\n"; exit 0
+
+    )
+
+  (* iterate over the graphs in previous state *)
+  let assign_fold_graphSet graph (lval, rval, ctx,set_acc) =
+    let lastNode = 
+      if (LocTraceGraph.is_empty graph)&&(match lval with (Var x, _) -> x.vglob | _ -> false) then
+        (
+          tidRecord#addTID 1;
+          {programPoint=ctx.prev_node;sigma=SigmaMap.empty;id=idGenerator#increment();tid=1;lockSet=VarinfoSet.empty}
+        )
+      else LocalTraces.get_last_node_progPoint graph ctx.prev_node
+    in
+    let new_set = 
+      (if Node.equal lastNode.programPoint LocalTraces.error_node then (print_string ("In assign, we have a trace that does not end in previous node:
+    "^(LocalTraces.show graph)^"\n"); set_acc) else 
+         let result_graphList = 
+           assign_on_node (if LocTraceGraph.is_empty graph then (LocTraceGraph.add_vertex graph lastNode) else graph) ctx lval rval lastNode
+         in
+         List.fold (fun set_fold result_graph -> if (LocalTraces.equal graph result_graph)  then set_acc else D.add result_graph set_fold) set_acc result_graphList
+      )
+    in
+    (lval, rval, ctx, new_set)
+
+  let assign ctx (lval:lval) (rval:exp) : D.t = 
+    predominatorRegistration#update ctx.prev_node ctx.node;
```

Comment:
While analyzing, a predominator analysis is conducted implicitely. With that, each edge effect updates the predominatorRegistration with an other edge that is registered.
In the future, it would be better to perform predominator analysis before beginning with the analysis